### PR TITLE
Handling a ROI when loading raster data

### LINKED ISSUE #645: [POC] Refactor CROP and load_rio()
# Context

Currently, the handling of region of interest (ROI) or crop geometry is split between _load_rio() and the crop() method. Which is hard to maintain. 
We propose to delete the ROI management from `_load_rio()` and use the `crop()` method before calling _load_rio(), if a ROI is provided during raster initialization.

### PR DESCRIPTION
**Pull request by vschaffn**
Resolves GlacioHack/xdem#602
Resolves #645

## Description
This PR introduces the functionality for handling a Region of Interest (ROI) in pixels when loading raster data.

## Changes
- Added `Raster.icrop()` method to handle ROI in pixels.

## Test
- Added `test_icrop()` in `test_raster` to ensure the good behavior when cropping a raster

## Doc
- Added `Raster.icrop()` along with `Raster.crop()` in doc, and making it clear that the crops are also functional when the Raster is not loaded into memory.

### CODE REVIEW

**Comment by rhugonnet** [comment]
Great addition, thanks @vschaffn! :slightly_smiling_face: 

I have a main remark about the `roi` addition to `_load_rio()` specifically:

As it happens, the `crop()` function already worked with the data not loaded on disk, by reading the ROI from a window in the same way you did (but it wasn't very clear from the documentation), see here: https://github.com/GlacioHack/geoutils/blob/514ea820dbfc660a0c2f498ebe75b41b0910667a/geoutils/raster/geotransformations.py#L496
In short, doing:
```python
rst = Raster("file.tif")
rst_cropped = rst.crop(my_roi)
```
worked without loading the full dataset.

We had been debating about adding the windowed reading of `crop` directly to `_load_rio()` to avoid the loading of `crop` being separate from all other loading of the `Raster` class. This would allow to use an `roi`/`crop_geom` from `__init__()` or `from_array()` directly, just as you did. 

Maybe this is the occasion to merge things together?
We could modify `_crop` to call your new `_load_rio(roi=)` (might require some small modifs following the code in `_crop`), and merge the `roi`/`crop_geom` arguments into one, to have everything consistent?

**Comment by vschaffn** [comment]
> the `crop()` function already worked with the data not loaded on disk

I had not noticed that the `crop()` method already worked with the data not loaded on disk. After studying `_crop()` function, I think the simplest thing is to restore `_load_rio()` without ROI, and to use crop, as it is, before `_load_rio()` if a ROI is passed when the raster is initialised. This will effectively mean that there is only one function dealing with the ROI, and therefore simplified code maintenance.

**Comment by vschaffn** [comment]
> Great addition, thanks @vschaffn! 🙂
> 
> I have a main remark about the `roi` addition to `_load_rio()` specifically:
> 
> As it happens, the `crop()` function already worked with the data not loaded on disk, by reading the ROI from a window in the same way you did (but it wasn't very clear from the documentation), see here:
> 
> https://github.com/GlacioHack/geoutils/blob/514ea820dbfc660a0c2f498ebe75b41b0910667a/geoutils/raster/geotransformations.py#L496
> 
> In short, doing:
> 
> ```python
> rst = Raster("file.tif")
> rst_cropped = rst.crop(my_roi)
> ```
> 
> worked without loading the full dataset.
> 
> We had been debating about adding the windowed reading of `crop` directly to `_load_rio()` to avoid the loading of `crop` being separate from all other loading of the `Raster` class. This would allow to use an `roi`/`crop_geom` from `__init__()` or `from_array()` directly, just as you did.
> 
> Maybe this is the occasion to merge things together? We could modify `_crop` to call your new `_load_rio(roi=)` (might require some small modifs following the code in `_crop`), and merge the `roi`/`crop_geom` arguments into one, to have everything consistent?

@rhugonnet I had not noticed that the crop() method already worked with the data not loaded on disk. After studying _crop() function, I think the simplest thing is to restore _load_rio() without ROI, and to use crop, as it is, before _load_rio() if a ROI is passed when the raster is initialised. This will effectively mean that there is only one function dealing with the ROI, and therefore simplified code maintenance.

**Comment by rhugonnet** [comment]
@vschaffn Yes that's true, this would work as well :slightly_smiling_face:. The difference would be only internal to make `_crop` depend on `_load_rio()`, so we don't need to do it (could at a later stage if we feel it is useful to avoid having independent loading calls, but that's not a priority).

**Comment by adebardo** [comment]
Okay for me, we can open an issue for the `_crop` dependance.

File: geoutils/raster/raster.py [hunk 1]
```diff
@@ -958,6 +1019,68 @@ def to_rio_dataset(self) -> rio.io.DatasetReader:
...
+        if not isinstance(transform, Affine):
+            transform = Affine(*transform)
+
+        roi_pix = {"x", "y", "w", "h"}
+        roi_georef = {"left", "bottom", "right", "top"}
+
+        # Check if ROI is georeferenced or pixel_based and convert to crs
+        if roi_georef.issubset(roi.keys()):
+            crs_roi = roi.get("crs", "EPSG:4326")
+            transformer = Transformer.from_crs(crs_roi, crs, always_xy=True)
```
Comment (adehecq) [inline geoutils/raster/raster.py:1051]: Note that we have a function to convert points from one CRS to another, which saves a few lines: https://github.com/GlacioHack/geoutils/blob/main/geoutils/projtools.py#L246

File: tests/test_raster/test_raster.py [hunk 2]
```diff
@@ -303,6 +312,41 @@ def test_load(self) -> None:
             r.filename = None
             r.load()
 
+    @pytest.mark.parametrize("example", [landsat_b4_path, aster_dem_path, landsat_rgb_path])  # type: ignore
+    def test_raster_with_roi(self, example: str) -> None:
```
[the remark this answers is not in the document]
Comment (adehecq) [reply tests/test_raster/test_raster.py:316]: Typically, we check that errors are raised properly, such as the ones raised at L 1075 in raster.py.

**Comment by adehecq** [comment]
Hi all,
Thanks for implementing this!

I'm slowly trying to catch up, and sorry if some of it should have come up earlier during the conception...
As raised by rhugonnet, it would have helped if you had known that crop was already working on the unloaded dataset (in general, we only load the data when strictly needed, i.e., for printing the data, changing the data or plotting, otherwise only metadata are loaded). Sorry, we could have clarified this earlier if we had been more available.
Considering this, I wonder if it would not be easier to update the method `crop` to take a geometry in pixels? Then either simply suggest people to use the crop method directly, or have an argument `crop` to Raster to enable direct cropping. In either case, I am a bit concerned that having an "ROI" functionality and "crop" functionality may create some confusion as both are somewhat redundant. What do you think @rhugonnet ?

One more thing to consider. We should make sure that crop and ROI (if kept separate) behave the same way when the ROI is partially or fully out of the raster bounds. I had to check with crop, but it looks like it does not raise any warning or error if the bound is larger (but silently crops to the maximum extent) and raises an error if there is no intersection at all between the two. See example below. Maybe we should raise a warning when the requested crop partially overlap and an error when it does not overlap at all? 
I made a comment directly in the code, but I think these edge cases should be tested in the tests, to make sure we don't end up with unexpected errors.

```
import geoutils as gu
img = gu.Raster(gu.examples.get_path("everest_landsat_b4"))
print(img.bounds)
# -> BoundingBox(left=478000.0, bottom=3088490.0, right=502000.0, top=3108140.0)

# Test with crop geometry too large on the right
img_crop = img.crop((478000, 3088490, 502100, 3108140), inplace=False)
print(img_crop.bounds)
# -> same extent as original

# Test with crop geometry outside of bounds 
img_crop = img.crop((502000, 3088490, 502500, 3108140), inplace=False)
# -> raises a rasterio WindowError -> we could actually catch this earlier and raise a clearer error 
```

**Comment by rhugonnet** [comment]
>  What do you think @rhugonnet ?

I agree, we should make the naming and behaviour fully consistent between `Raster(roi=)` and `Raster.crop(crop_geom=)`, as they are the same functionality.
I'd suggest simply renaming `roi` into `crop_geom` in `Raster.__init__`. Then, as @vschaffn already proposed, calling `crop(crop_geom=)` directly. This will ensure everything is the same, and rely on all existing tests to reduce maintenance of this addition :wink: .

**Comment by adebardo** [comment]
image

**Review by adebardo — COMMENTED** [review]
I think we can put the copyright CNES.

**Review by adebardo — COMMENTED** [review]
THE CI is failing because of the documentation

**Comment by vschaffn** [comment]
> THE CI is failing because of the documentation

fixed

**Comment by vschaffn** [comment]
@adebardo I fixed the documentation issue in the CI, and I added CNES copyright to `raster.py` file

**Review by adebardo — APPROVED** [review]
